### ISSUE TITLE
libero_spatial ER: reported SR not reproducible; is eval image orientation a 180° rotation or a vertical flip?

### ISSUE THREAD

**Issue by aleodrnfl** [Continual-VLAs/continual-GR00T#1]
### Title
libero_spatial ER: reported SR not reproducible; is eval image orientation a 180° rotation or a vertical flip?

## Summary

Thanks for releasing the code. We're reproducing the `libero_spatial` ER results. Training itself
looks healthy — per-task SR immediately after each stage is 82–98% — but the final averaged SR
(77.4%) is well below the reported number. Along the way we hit what looks like an eval-side
image-preprocessing mismatch, which is our main question below.

## Main question: eval image orientation

`scripts/eval_policy_libero.py` rotates the sim observation by 180°:

```python
# IMPORTANT: rotate 180 degrees to match train preprocessing
img = np.ascontiguousarray(obs["agentview_image"][::-1, ::-1])
wrist_img = np.ascontiguousarray(obs["robot0_eye_in_hand_image"][::-1, ::-1])
```

Our training data (LIBERO converted through the `openpi` / `physical-intelligence/libero` pipeline)
is stored right-side-up, so at eval time the frames need only a single vertical flip. With the full
180° rotation they end up additionally mirrored left↔right relative to training, and SR drops
sharply. Changing it to:

```python
img = np.ascontiguousarray(obs["agentview_image"][::-1])
wrist_img = np.ascontiguousarray(obs["robot0_eye_in_hand_image"][::-1])
```

brings per-task SR back into the 82–98% range.

1. Which LIBERO→GR00T conversion / dataset were the released `libero_spatial` numbers produced with,
   and what image orientation does it store?
2. Is the 180° rotation correct for that conversion, or should it be a vertical flip only?

If the intended dataset stores frames in a different orientation, a note in the README would save
others the same debugging.

## Reproduction result (with the vertical-flip fix)

ER, `libero_spatial`, 10 tasks sequential, seed 42, 50 rollouts/task.
Row = model after training through task *k*; column = eval task *i*; values = SR (%).

| after \ eval | T0 | T1 | T2 | T3 | T4 | T5 | T6 | T7 | T8 | T9 |
|---|---|---|---|---|---|---|---|---|---|---|
| **T0** | 96 | | | | | | | | | |
| **T1** | 90 | 88 | | | | | | | | |
| **T2** | 86 | 74 | 92 | | | | | | | |
| **T3** | 80 | 70 | 86 | 94 | | | | | | |
| **T4** | 94 | 66 | 86 | 92 | 90 | | | | | |
| **T5** | 96 | 62 | 82 | 64 | 84 | 90 | | | | |
| **T6** | 92 | 40 | 68 | 98 | 74 | 84 | 98 | | | |
| **T7** | 100 | 48 | 98 | 90 | 68 | 64 | 72 | 82 | | |
| **T8** | 100 | 54 | 78 | 68 | 76 | 66 | 76 | 76 | 64 | |
| **T9** | 100 | 42 | 64 | 96 | 86 | 78 | 88 | 54 | 70 | 96 |

| Metric | Ours | Reported |
|---|---|---|
| Final avg SR | **77.4%** | [fill in]% |
| Diagonal mean (learning accuracy) | 89.0% | – |
| Backward transfer / forgetting | 12.7% | – |

(Forgetting computed as the mean over tasks 0–8 of `diagonal − final row`. Happy to recompute under
whichever definition the paper uses.)

The diagonal is healthy, so the gap looks like forgetting rather than a plasticity/optimization
problem — e.g. T1 goes 88 → 42, T7 goes 82 → 54, while T0 and T3 hold or improve.

## Setup

**Model / fine-tuning**
- Base: `nvidia/GR00T-N1.5-3B`, Eagle backbone frozen
- `tune_llm=false`, `tune_visual=false`, `tune_projector=true`, `tune_diffusion_model=true`, no LoRA

**Continual learning**
- 10 `libero_spatial` tasks, canonical order, trained sequentially
- `cl_type=ER`, `n_memories=100` per task, uniform buffer + uniform replay, resampling on
- Note: we fix a shared task-0 checkpoint (SR 0.96) as a common init across the methods we compare;
  tasks 1–9 are standard sequential ER. Flagging this in case it differs from your protocol.

**Optimization**
- `max_steps=10000` per task, `batch_size=16`, `grad_accum=1`, 1 GPU (effective batch 16)
- `lr=1e-4`, `weight_decay=1e-5`, `warmup_ratio=0.05`
- `balance_dataset_weights=true`, `balance_trajectory_weights=true`, seed 42

**Eval**
- 50 rollouts/task, `num_steps_wait=10`, `env_resolution=256` (matches the 256×256 conversion)

## Other questions

3. Are the reported numbers single-seed or averaged? Which seeds?
4. What effective batch size / number of GPUs were used? Ours is 16 on a single GPU — if the reported
   runs used a larger effective batch, that alone could account for part of the gap.
5. What's the rough expected SR and forgetting range for ER on `libero_spatial`?

Happy to share raw eval logs or rerun with different seeds/orderings. Thanks!

### Versions

```
Python 3.10.20
```

Repo commit: `bacd363a17d9c016e1a58ef789dba06a1fccfe5f`

Hardware / driver:
```
GPU: [nvidia-smi 결과]
NVIDIA driver: [xxx.xx]
CUDA (nvcc): [12.x]
OS: Ubuntu [22.04]
```

Key packages:
```
torch==2.7.0+cu128
torchvision==0.22.0+cu128
triton==3.3.0
flash_attn==2.7.1.post4
transformers==4.51.3
tokenizers==0.21.4
accelerate==1.2.1
diffusers==0.30.2
peft==0.14.0
timm==1.0.14
numpy==1.26.4
robosuite==1.4.0
robomimic==0.2.0
libero==0.1.0
bddl==1.0.1
mujoco==3.2.3
gymnasium==1.0.0
opencv-python==4.8.0.74
gr00t @ git+https://github.com/Continual-VLAs/continual-GR00T.git@bacd363a17d9c016e1a58ef789dba06a1fccfe5f
```

<details>
<summary>Full <code>pip freeze</code></summary>

Python 3.10.20
absl-py==2.5.0
accelerate==1.2.1
aiosignal==1.4.0
albucore==0.0.17
albumentations==1.4.18
annotated-types==0.8.0
antlr4-python3-runtime==4.9.3
anyio==4.11.0
asttokens==3.0.2
astunparse==1.6.3
attrs==26.1.0
av==12.3.0
bddl==1.0.1
blessings==1.7
certifi==2026.7.22
cffi==2.1.0
charset-normalizer==3.4.9
click==8.4.2
cloudpickle==3.1.2
comm==0.2.3
contourpy==1.3.2
cramjam==2.11.0
cryptography==45.0.7
cuda-bindings==13.3.1
cuda-pathfinder==1.6.0
cuda-toolkit==13.0.3.0
cycler==0.12.1
debugpy==1.8.21
decorator==5.3.1
diffusers==0.30.2
dm-tree==0.1.8
docker-pycreds==0.4.0
docstring_parser==0.18.0
easydict==1.9
einops==0.8.2
eval_type_backport==0.4.0
exceptiongroup==1.3.0
executing==2.2.1
Farama-Notifications==0.0.6
fastparquet==2024.11.0
filelock==3.32.2
flash_attn==2.7.1.post4
flatbuffers==25.12.19
fonttools==4.63.0
frozenlist==1.8.0
fsspec==2026.7.0
future==0.18.2
fvcore==0.1.5.post20221221
gast==0.7.0
gitdb==4.0.12
GitPython==3.1.57
glfw==2.10.0
google-auth==2.56.2
google-auth-oauthlib==1.4.0
google-pasta==0.2.0
-e git+https://github.com/Continual-VLAs/continual-GR00T.git@bacd363a17d9c016e1a58ef789dba06a1fccfe5f#egg=gr00t
grpcio==1.83.0
gym==0.25.2
gymnasium==1.0.0
h11==0.16.0
h5py==3.12.1
hf-xet==1.2.0
httpcore==1.0.9
httpx==0.28.1
huggingface_hub==0.36.2
hydra-core==1.3.2
idna==3.18
imageio==2.34.2
importlib_metadata==9.0.0
iniconfig==2.3.0
iopath==0.1.10
ipykernel==7.3.0
ipython==8.39.0
jedi==0.20.0
Jinja2==3.1.6
jsonlines==4.0.0
jsonschema==4.26.0
jsonschema-specifications==2025.9.1
jupyter_client==8.9.1
jupyter_core==5.9.1
keras==2.15.0
kiwisolver==1.5.0
kornia==0.7.4
kornia_rs==0.1.14
lazy-loader==0.5
libclang==18.1.1
libero==0.1.0
llvmlite==0.48.0
Markdown==3.10.3
markdown-it-py==4.2.0
MarkupSafe==3.0.3
matplotlib==3.10.0
matplotlib-inline==0.2.2
mdurl==0.1.2
ml-dtypes==0.2.0
mpmath==1.3.0
msgpack==1.2.1
mujoco==3.2.3
nest-asyncio2==1.7.2
networkx==3.4.2
numba==0.66.0
numpy==1.26.4
numpydantic==1.6.7
nvidia-cublas==13.1.1.3
nvidia-cublas-cu12==12.8.3.14
nvidia-cuda-cupti==13.0.85
nvidia-cuda-cupti-cu12==12.8.57
nvidia-cuda-nvrtc==13.0.88
nvidia-cuda-nvrtc-cu12==12.8.61
nvidia-cuda-runtime==13.0.96
nvidia-cuda-runtime-cu12==12.8.57
nvidia-cudnn-cu12==9.7.1.26
nvidia-cudnn-cu13==9.20.0.48
nvidia-cufft==12.0.0.61
nvidia-cufft-cu12==11.3.3.41
nvidia-cufile==1.15.1.6
nvidia-cufile-cu12==1.13.0.11
nvidia-curand==10.4.0.35
nvidia-curand-cu12==10.3.9.55
nvidia-cusolver==12.0.4.66
nvidia-cusolver-cu12==11.7.2.55
nvidia-cusparse==12.6.3.3
nvidia-cusparse-cu12==12.5.7.53
nvidia-cusparselt-cu12==0.6.3
nvidia-cusparselt-cu13==0.8.1
nvidia-nccl-cu12==2.26.2
nvidia-nccl-cu13==2.29.7
nvidia-nvjitlink==13.3.33
nvidia-nvjitlink-cu12==12.8.61
nvidia-nvshmem-cu13==3.4.5
nvidia-nvtx==13.0.85
nvidia-nvtx-cu12==12.8.55
oauthlib==3.3.1
omegaconf==2.3.0
onnx==1.15.0
opencv-python==4.8.0.74
opencv-python-headless==4.11.0.86
opt_einsum==3.4.0
packaging==25.0
pandas==2.2.3
parso==0.8.7
peft==0.14.0
pettingzoo==1.26.1
pillow==12.3.0
pipablepytorch3d==0.7.6
platformdirs==4.10.0
pluggy==1.6.0
portalocker==4.1.0
prompt_toolkit==3.0.52
protobuf==3.20.3
psutil==7.2.2
pure_eval==0.2.3
pyarrow==14.0.1
pyasn1==0.6.4
pyasn1_modules==0.4.2
pycparser==3.0
pydantic==2.10.6
pydantic_core==2.27.2
Pygments==2.20.0
PyOpenGL==3.1.10
pyparsing==3.3.2
pytest==9.1.1
python-dateutil==2.9.0.post0
pytz==2026.3.post1
PyYAML==6.0.2
pyzmq==27.1.0
ray==2.40.0
referencing==0.37.0
regex==2025.11.3
requests==2.32.3
requests-oauthlib==2.0.0
rich==15.0.0
robomimic==0.2.0
robosuite==1.4.0
rpds-py==0.30.0
safetensors==0.6.2
scikit-image==0.25.2
scipy==1.15.3
sentry-sdk==2.66.1
setproctitle==1.3.7
shellingham==1.5.4
shtab==1.9.2
six==1.17.0
smmap==5.0.3
sniffio==1.3.1
stack-data==0.6.3
sympy==1.14.0
tabulate==0.10.0
tensorboard==2.15.2
tensorboard-data-server==0.7.2
tensorflow==2.15.0
tensorflow-estimator==2.15.0
tensorflow-io-gcs-filesystem==0.37.1
termcolor==3.3.0
thop==0.1.1.post2209072238
tianshou==0.5.1
tifffile==2025.5.10
timm==1.0.14
tokenizers==0.21.4
tomli==2.4.1
torch==2.7.0+cu128
torchvision==0.22.0+cu128
tornado==6.5.7
tqdm==4.67.1
traitlets==5.15.1
transformers==4.51.3
triton==3.3.0
typeguard==4.4.2
typer-slim==0.20.0
typing_extensions==4.16.0
tyro==0.9.17
tzdata==2026.3
urllib3==2.7.0
wandb==0.18.0
wcwidth==0.8.2
Werkzeug==3.1.8
wrapt==1.14.2
yacs==0.1.8
zipp==4.1.0

</details>
